### PR TITLE
Weakened off what "static linking" means when building for OS X.

### PR DESCRIPTION
**Pull request by bobsummerwill**
Apple explicitly does not support statically linked binaries (i.e., for the executable itself) on OS X.
See https://developer.apple.com/library/mac/qa/qa1118/_index.html for more info.
But it does support static linkage of the external libraries being used by that executable, which is almost as good.
This is part of https://github.com/ethereum/webthree-umbrella/issues/495, where we are adding support for static builds to the codebase.

For OS X we currently only have partially-static library linkage too.   We would need to build the following from source to have .a files to link against instead of their dylibs which we get from Homebrew:
- jsoncpp
- json-rpc-cpp
- leveldb
- miniupnp
- gmp

Two further libraries (curl and zlib) ship as dylibs with the platform but again we could build from source and statically link these too.

RelWithDebugInfo for a statically linked OS X binary is 13MB.

### CODE REVIEW

**Comment by bobsummerwill** [comment]
CC @rainbeam.

File: cmake/EthDependencies.cmake [hunk 1]
```diff
@@ -89,7 +82,7 @@ See https://developer.apple.com/library/mac/qa/qa1118/_index.html for more info.
 	if(MSVC)
 		set(CMAKE_FIND_LIBRARY_SUFFIXES .lib .a ${CMAKE_FIND_LIBRARY_SUFFIXES})
 	else()
-		set(CMAKE_FIND_LIBRARY_SUFFIXES .a )
+		set(CMAKE_FIND_LIBRARY_SUFFIXES .a .dylib)
```
Comment (rainbreak) [inline cmake/EthDependencies.cmake:103]: Add a comment connecting .dylib to osx.

File: cmake/EthExecutableHelper.cmake [hunk 2]
```diff
@@ -46,10 +46,14 @@ endmacro()
...
 	add_executable(${EXECUTABLE} ${SRC_LIST} ${HEADERS})
 
-	if (STATIC_LINKING)
-		set(CMAKE_EXE_LINKER_FLAGS "-static ${CMAKE_EXE_LINKER_FLAGS}")
-		set_target_properties(${EXECUTABLE} PROPERTIES LINK_SEARCH_START_STATIC 1)
-		set_target_properties(${EXECUTABLE} PROPERTIES LINK_SEARCH_END_STATIC 1)
+	# Apple does not support statically linked binaries on OS X.
+	# See https://developer.apple.com/library/mac/qa/qa1118/_index.html for more info.
+	if (NOT APPLE)
+		if (STATIC_LINKING)
```
Comment (rainbreak) [inline cmake/EthExecutableHelper.cmake]: `if (STATIC_LINKING AND NOT APPLE)` would scan better and avoid nested `if`

File: cmake/EthDependencies.cmake (hunk 1 above)
Comment (rainbreak) [inline cmake/EthDependencies.cmake:103]: Not sure how the precedence works here. Will it prioritise .a over .dylib where both are present?

**Comment by rainbreak** [comment]
Makes sense. Work with what the platform gives you. I really don't know much about OSX though so no idea on distribution implications of static linking.

Is that 13MB binary stripped?

File: cmake/EthDependencies.cmake (hunk 1 above)
Comment (bobsummerwill) [inline cmake/EthDependencies.cmake:103]: It appears to do that, yes.

**Comment by bobsummerwill** [comment]
Updated as per comments.   Got OK from Greg with those caveats.   Merging.

**Comment by bobsummerwill** [comment]
@rainbeam - "Is that 13MB binary stripped?"

I think so, though I am not 100% sure.  It's certainly massively smaller than it was, but I suspect it will go smaller again when we have all the libraries statically linked.     I'll really dig into the linker settings for static linking when all these other more immediate issues are resolved.

Trying to do WIndows static build now.

**Pull request merged**